### ISSUE TITLE
feat(retrieval): reranker モデルを bge-reranker-base に更新して Static NC 改善

### ISSUE THREAD

**Issue by Kewton** [Kewton/photon-mlx#89]
## 背景

Epic #81 の Sub-Issue。現在の reranker は `cross-encoder/ms-marco-MiniLM-L-6-v2`（一般 IR tuned、英語 MS MARCO で学習された 22M cross-encoder）。`baseline_reporag/retrieval/reranker.py` のコメントにも「trained on English MS MARCO data」と明記されており、日本語クエリでは性能劣化が見込まれるため英語 expansion term を rerank_query として使用している。より新しく多言語 fine-tune された reranker に差し替えることで Static NC 改善を狙う。

> 親 Epic 目標との関係: Epic #81 全体目標は Static NC < 15%（-1.7pp）。本 Issue 単体の目標は < 16%（-0.7pp）で、残りの差分は他 Sub-Issue（retrieval パラメータ最適化、embedding モデル更新等）で埋める前提。

### 最有力: `BAAI/bge-reranker-base`
- パラメータ: 278M（HuggingFace モデルカード記載値、実装時に実測確認）
- 多言語 IR タスクで fine-tuned されており、日本語混在クエリでの性能劣化が小さいことが期待される（コード領域での実性能は本 Issue のベンチマークで検証する）
- IR benchmark（BEIR 等）では MS-MARCO MiniLM を上回る報告があるが、本 RepoRAG タスクでの有効性は未確認 → 要実測
- sentence-transformers の `CrossEncoder` API でロード可能（公式 FlagEmbedding リポで互換性を明記）

### 次点候補
- `BAAI/bge-reranker-v2-m3`（多言語特化、より重い 568M）
- `Alibaba-NLP/gte-multilingual-reranker-base`（日本語も強い）

### 1. Reranker モデルの差し替え（YAML + ハードコード default の同時更新）

```yaml
# configs/baseline.yaml / configs/photon_small.yaml / configs/photon_long_context.yaml
retrieval:
  reranker:
    enabled: true
    model_id: "BAAI/bge-reranker-base"  # was: cross-encoder/ms-marco-MiniLM-L-6-v2
```

注 1: `configs/photon_tiny.yaml` / `configs/photon_tiny_recgen.yaml` / `configs/photon_600m_paper.yaml` は `retrieval.reranker` セクションを持たないため変更対象外。

注 2: 以下 3 箇所の**ハードコード default** も同時に更新する（YAML 未指定時の fallback が旧モデルに残る後方互換の罠を防ぐため）：

- `baseline_reporag/retrieval/reranker.py` コンストラクタ引数 default（~L50）
- `baseline_reporag/pipeline_factory.py` `reranker_cfg.get("model_id", ...)` fallback（~L128）
- `app/photon_app.py` 同 fallback（~L1025）

### 2. CrossEncoderReranker の動作確認

現行 `baseline_reporag/retrieval/reranker.py` は `sentence_transformers.CrossEncoder.predict(pairs)` を使用する PyTorch ベース実装（MLX には非依存）。以下を smoke test で検証する：

- `CrossEncoder("BAAI/bge-reranker-base")` がロード可能なこと
- `predict(list[tuple[str, str]])` が `numpy.ndarray` 形状 `(N,)` の浮動小数スコアを返すこと
- **出力スコアスケール**の記録（min/max/mean）。現行 MiniLM は logit (-10～+10 前後)、bge-reranker 系はモデルによりスケールが異なる可能性あり。`file_type_boost` を将来有効化した際に相対関係が壊れないかを確認し、必要なら adapter 層で sigmoid 正規化を挟む
- noise filter（`_NOISE_PATTERNS`）が従来どおり前段で動作すること
- rerank_query 優先ロジックが維持されること

### 3. max_length の設計判断

現行実装は `max_length=256` 固定（`reranker.py:45-51`）。bge-reranker-base は 512 トークンまで対応するため、以下のいずれかを採用する：

- **A. 256 据え置き（推奨）**: 既存の `content[:600]` 文字トリム（`reranker.py:90`）と整合し、レイテンシへの影響最小。
- **B. 512 に拡張**: 長いコードチャンクでの relevance 精度向上が期待できるが、CrossEncoder 内部計算量が概ね 4 倍に増える可能性あり。

設計方針書で選択し、選択した設定でベンチマーク計測を実施する。

### 4. `rerank_query` 使い分けポリシーの決定

現行 `baseline_reporag/pipeline.py:161` は日本語クエリ時に英語 expansion term を `rerank_query` として渡して cross-encoder の英語特化を回避している。bge-reranker-base は多言語対応のため、以下のどちらが Static NC に有利かを検証する：

- **(a) 従来どおり**: 英語 expansion を `rerank_query` として渡す
- **(b) 日本語クエリそのまま**: `rerank_query` を指定せず `query` を使う

最低限、設計方針書でどちらを採用するか明記する。時間があれば A/B 計測して良い方を採用。

### 5. 単体テストの追加

`baseline_reporag/tests/test_reranker.py` を新規作成し、以下の基本挙動を検証する：

- noise filter による既知ノイズファイル除去
- 空入力 / `top_k` 切り取り
- `rerank_query` 優先ロジック
- （optional / pytest marker 付き）実モデル `BAAI/bge-reranker-base` のロード成功と `predict()` 戻り値形状

実モデルロードを伴うテストは `@pytest.mark.slow` など marker で gate し、CI では skip（または事前キャッシュ前提で実行）する。

### 6. ベンチマーク & ドキュメント更新

- Static NC 120 Q（現行 MiniLM baseline の再計測値 vs. bge-reranker）
- レイテンシ比較（**warmup 後の P50**、cold start ロード時間は除外）
- bench/ ハーネスを用いた実行コマンドは work-plan で具体化
- `reports/bge_reranker_eval.md` の冒頭に「本レポート以降の値は bge-reranker-base 採用後。以前のレポート（gate2_judgment_v*.md 等）は MiniLM 前提」の注記を追加

### コード
- `baseline_reporag/retrieval/reranker.py`（default model_id 更新 / API 差異があれば adapter 層追加 / max_length 判断反映 / docstring のモデル名更新）
- `baseline_reporag/pipeline_factory.py`（`reranker_cfg.get("model_id", ...)` fallback 更新）
- `app/photon_app.py`（同 fallback 更新）
- `baseline_reporag/tests/test_reranker.py`（新規）

### 設定
- `configs/baseline.yaml`（`retrieval.reranker.model_id` 更新）
- `configs/photon_small.yaml`（同）
- `configs/photon_long_context.yaml`（同）

### ドキュメント
- `docs/deployment.md`（model_id 表、ダウンロードサイズ記述を更新: ~100MB → ~550MB 目安）
- `docs/troubleshooting.md`（キャッシュディレクトリパス更新: `models--cross-encoder--ms-marco-...` → `models--BAAI--bge-reranker-base`）
- `docs/tutorial.md`（ダウンロードサイズ・reranker 言及箇所更新）
- `reports/bge_reranker_eval.md`（新規 / 実測結果 + 過去レポートとの連続性注記）

## 受入条件

- [ ] `BAAI/bge-reranker-base` が sentence-transformers の `CrossEncoder` API で動作（smoke test パス）
- [ ] `baseline_reporag/tests/test_reranker.py` を新規追加し、CrossEncoderReranker の基本挙動回帰を保護
- [ ] `CrossEncoderReranker.rerank(query, results, store, top_k, rerank_query)` の署名・戻り値型 (`list[RetrievalResult]`) の**後方互換を維持**し、既存 mock ベーステスト（test_photon_pipeline.py / test_pipeline_integration.py）が**無改変で pass** すること
- [ ] Static NC が現在の 16.7% から改善、**本 Issue 単体の目標 < 16%**（Epic #81 全体目標 < 15% の部分貢献）
- [ ] レイテンシ悪化が **warmup 後の Static P50 で +3s 以内**（cold start 除外）
- [ ] MT NC 5.6% を維持（reranker は follow-up turn ではスキップされる仕様のため、MT NC への影響は Turn 1 に限定される）
- [ ] docs/ の 3 本（deployment / troubleshooting / tutorial）で旧モデル名・サイズ記述が残らないこと
- [ ] `ruff check` / `pytest` 全パス

## 関連

- 親 Epic: #81

**Comment by Kewton** [comment]
## Abort — Retry 測定結果

Wave 6 全 revert 後、単独 PR として #89 を再着手し pre-merge Static NC eval を実行。結果として **bge-reranker-base は本 workload では Static NC を改善せず、むしろ悪化**することが確認されました。

### 実測

| Reranker | Static NC (120Q) | Delta |
|----------|------------------|-------|
| `cross-encoder/ms-marco-MiniLM-L-6-v2` (current) | **17.5%** | — |
| `BAAI/bge-reranker-base` (#89 target) | **18.3%** | **+0.8pp 悪化** |

### 判断根拠

1. Issue #89 目標 (< 16%) 未達かつ方向が regression
2. Wave 5 分析（`reports/two_pass_search_analysis.md`）で既に `MS-MARCO MiniLM > bge` と結論済み
3. Wave 6 全 revert 時の MT regression 要因推定と整合

### 結論

- 本 Issue は **No-Go でクローズ**
- Epic #81 (Static NC < 15%) は他 Sub-Issue（#88 grid search、#90 embedding 更新）で進める

詳細レポート: `reports/issue_89_bge_reranker_retry_abort.md`
